### PR TITLE
Makes automatic optimization a model attribute

### LINKED ISSUE #4317: Automatic Optimization should be a property of the LightningModule, not the Trainer
## 🚀 Feature
https://pytorch-lightning.readthedocs.io/en/latest/optimizers.html#manual-optimization
To switch between manual and automatic optimization, we need to toggle this flag on the `Trainer`. However, all of the user code for optimization is inside the `LightningModule`. If we had this flag for `automatic_optimization` on the LightningModule instead, we'd keep all the user-facing optimization pieces in one place.

### Pitch
Define a property method for `automatic_optimization` inside the LightningModule. this would be another model hook

By default:
```
@property
def automatic_optimization(self) -> bool:
    return True
```

Then we can set the corresponding properties on the trainer/training loop based on the `LightningModule` and deprecate the `automatic_optimization` flag on the Trainer

### PR DESCRIPTION
**Pull request by justusschock**
## What does this PR do?

Makes automatic optimisation a model property and deprecates the trainer flag. Fixes #4317

## Before submitting
- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? 
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In in short, see following bullet-list:  

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified; _Bugfixes should be including in bug-fix release milestones (m.f.X) and features should be included in (m.X.b) releases._
 

## Did you have fun?
Make sure you had fun coding 🙃

cc @ananthsub

### CODE REVIEW

**Review by tchaton — APPROVED** [review]

**Review by SeanNaren — APPROVED** [review]

**Comment by SeanNaren** [comment]
@justusschock able to fix the failing codefactor test?

**Review by ydcjeff — APPROVED** [review]

File: pytorch_lightning/trainer/connectors/model_connector.py [hunk 1]
```diff
@@ -35,6 +35,9 @@ def copy_trainer_model_properties(self, model):
         else:
             ref_model = model
 
+        automatic_optimization = ref_model.automatic_optimization and self.trainer.train_loop.automatic_optimization
+        self.trainer.train_loop.automatic_optimization = automatic_optimization
```
Comment (ananthsub) [inline pytorch_lightning/trainer/connectors/model_connector.py:39]: should this be `or` for backwards compatibility? after this PR is merged, someone could set Trainer(automatic_optimization=False) but that would be overridden by the `automatic_optimization` set on the lightning module right?

File: pytorch_lightning/trainer/trainer.py [hunk 2]
```diff
@@ -346,6 +347,14 @@ def __init__(
         )
 
         # init train loop related flags
+        # TODO: deprecate in 1.2.0
+        if automatic_optimization is None:
+            automatic_optimization = True
```
Comment (ananthsub) [inline pytorch_lightning/trainer/trainer.py:352]: should the typehint above be changed to Optional[bool] ?

File: pytorch_lightning/core/lightning.py [hunk 3]
```diff
@@ -159,6 +159,14 @@ def on_gpu(self):
         """
         return self.device.type == "cuda"
 
+    @property
+    def automatic_optimization(self) -> bool:
+        """
+        If False you are responsible for calling .backward, .step, zero_grad.
+        Meant to be used with multiple optimizers by advanced users.
```
Comment (rohitgr7) [inline pytorch_lightning/core/lightning.py:166]: I think the above line is sufficient. A noob user can use it with a single optimizer too.
```suggestion
```

File: pytorch_lightning/trainer/trainer.py [hunk 4]
```diff
@@ -208,7 +208,8 @@ def __init__(
             log_every_n_steps: How often to log within steps (defaults to every 50 steps).
 
             automatic_optimization: If False you are responsible for calling .backward, .step, zero_grad.
-                Meant to be used with multiple optimizers by advanced users.
+                Meant to be used with multiple optimizers by advanced users. Passing bool
+                value is deprecated in v1.1 and will be removed in v1.3.
```
Comment (rohitgr7) [inline pytorch_lightning/trainer/trainer.py:212]:
```suggestion
                If False you are responsible for calling .backward, .step, zero_grad in LightningModule.
                This argument has been moved to LightningModule. It is deprecated here in v1.1 and
                will be removed in v1.3.
```

**Review by rohitgr7 — APPROVED** [review]
Just some comments. Else LGTM.

**Review by s-rog — APPROVED** [review]

File: pytorch_lightning/trainer/connectors/model_connector.py (hunk 1 above)
Comment (justusschock) [reply pytorch_lightning/trainer/connectors/model_connector.py:39]: The current idea is that the default is False and you can turn it off by either the module or the trainer flag. So if one of them resolves to false, the result should be false as well.

**Review by ananthsub — APPROVED** [review]
LGTM!

**Comment by ananthsub** [comment]
@justusschock we need a corresponding setter decorator to actually change this on the module easily. i can send a quick PR
